### PR TITLE
Linter updates: enable "bugs" preset

### PR DESCRIPTION
**Pull request by Porges**
Enable the "`bugs`" preset: the important changes here (IMO) are that `gosec` and `exhaustive` get added. I removed the explicit inclusion of `exportloopref` as it is part of `bugs`, and also added `megacheck` to retain the `gosimple` linter which for some reason gets dropped otherwise with these changes.

Diff of output of `golangci-lint linters`:
 ```
Enabled by your configuration linters:
+asciicheck: Simple linter to check that your code does not contain non-ASCII identifiers [fast: true, auto-fix: false]
+bodyclose: checks whether HTTP response body is closed successfully [fast: true, auto-fix: false]
 deadcode: Finds unused code [fast: true, auto-fix: false]
 errcheck: Errcheck is a program for checking for unchecked errors in go programs. These unchecked errors can be critical bugs in some cases [fast: true, auto-fix: false]
+exhaustive: check exhaustiveness of enum switch statements [fast: true, auto-fix: false]
 exportloopref: checks for pointers to enclosing loop variables [fast: true, auto-fix: false]
 gofmt: Gofmt checks whether code was gofmt-ed. By default this tool runs with -s option to check for code simplification [fast: true, auto-fix: true]
+gosec (gas): Inspects source code for security problems [fast: true, auto-fix: false]
 gosimple (megacheck): Linter for Go source code that specializes in simplifying a code [fast: true, auto-fix: false]
 govet (vet, vetshadow): Vet examines Go source code and reports suspicious constructs, such as Printf calls whose arguments do not align with the format string [fast: true, auto-fix: false]
 ineffassign: Detects when assignments to existing variables are not used [fast: true, auto-fix: false]
+noctx: noctx finds sending http request without context.Context [fast: true, auto-fix: false]
+rowserrcheck: checks whether Err of rows is checked successfully [fast: true, auto-fix: false]
+sqlclosecheck: Checks that sql.Rows and sql.Stmt are closed. [fast: true, auto-fix: false]
 staticcheck (megacheck): Staticcheck is a go vet on steroids, applying a ton of static analysis checks [fast: true, auto-fix: false]
 structcheck: Finds unused struct fields [fast: true, auto-fix: false]
 typecheck: Like the front-end of a Go compiler, parses and type-checks Go code [fast: true, auto-fix: false]
+unparam: Reports unused function parameters [fast: true, auto-fix: false]
 unused (megacheck): Checks Go code for unused constants, variables, functions and types [fast: false, auto-fix: false]
 varcheck: Finds unused global variables and constants [fast: true, auto-fix: false]
```

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -4,4 +4,9 @@ run:
 linters:
   enable:
     - gofmt
-    - exportloopref
+    - megacheck
+  presets:
+    - bugs
+    - unused
+  disable: 
+    - scopelint # obsoleted by exportloopref
```
Comment (matthchr) [inline .golangci.yml:12]: You removed exportloopref though?

Comment (Porges) [reply .golangci.yml:12]: it's part of the `bugs` preset, see diff in the PR comment

Comment (matthchr) [reply .golangci.yml:12]: Ah, nice -- I had missed that in the clutter

Comment (Porges) [reply .golangci.yml:12]: Reworded PR comment to make this clearer 👍

File: hack/generator/pkg/astmodel/package_definition.go [hunk 2]
```diff
@@ -175,7 +175,7 @@ func emitGroupVersionFile(pkgDef *PackageDefinition, outputDir string) error {
 
 	gvFile := filepath.Join(outputDir, "groupversion_info"+CodeGeneratedFileSuffix)
 
-	err = ioutil.WriteFile(gvFile, buf.Bytes(), 0700)
+	err = ioutil.WriteFile(gvFile, buf.Bytes(), 0600)
```
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/package_definition.go:178]: What does this magic number mean?

File: controllers/generic_controller.go [hunk 3]
```diff
@@ -379,16 +379,15 @@ func (gr *GenericReconciler) reconcileDelete(ctx context.Context, metaObj azcore
...
-	case zips.DeletingProvisioningState:
+	if resource.ProvisioningState == zips.DeletingProvisioningState {
 		msg := "deleting... checking for updated state"
 		gr.Recorder.Event(metaObj, v1.EventTypeNormal, "ResourceDeleteInProgress", msg)
 		return gr.updateFromNonTerminalDeleteState(ctx, resource, metaObj)
-	default:
-		msg := fmt.Sprintf("start deleting resource in state %q", resource.ProvisioningState)
-		gr.Recorder.Event(metaObj, v1.EventTypeNormal, "ResourceDeleteStart", msg)
-		return gr.startDeleteOfResource(ctx, resource, metaObj)
 	}
```
Comment (theunrepentantgeek) [inline controllers/generic_controller.go:386]: Interesting that the linter objects to simple **switch** statements.

File: hack/generator/pkg/astmodel/package_definition.go (hunk 2 above)
Comment (Porges) [reply hack/generator/pkg/astmodel/package_definition.go:178]: `rw` for owner, nothing for anyone else (previously it was `rwx`, which `gosec` objects to)

File: controllers/generic_controller.go (hunk 3 above)
Comment (Porges) [reply controllers/generic_controller.go:386]: it objects to non-exhaustive switches. in this case it was simpler to convert it to `if-else`

File: hack/generator/pkg/astmodel/package_definition.go (hunk 2 above)
Comment (Porges) [reply hack/generator/pkg/astmodel/package_definition.go:178]: obviously having constants defined in the stdlib is too ivory-tower 😉

**Comment by Porges** [comment]
Will await sign-off from @devigned

**Comment by matthchr** [comment]
Merging this so that I can fix any issues that it raises in my open PR

**Pull request merged**